examples/darwin-framework-tool/commands/common/ControllerStorage.mm

```diff
@@ -0,0 +1,132 @@
+    if (data == nil) {
+        return data;
+    }
+
+    NSError * error;
+    id value = [NSKeyedUnarchiver unarchivedObjectOfClasses:MTRDeviceControllerStorageClasses() fromData:data error:&error];
+    return value;
+}
+
+- (BOOL)controller:(MTRDeviceController *)controller
+        storeValue:(id<NSSecureCoding>)value
+            forKey:(NSString *)key
+     securityLevel:(MTRStorageSecurityLevel)securityLevel
+       sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Set value for %@", controller, key);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+    NSError * error;
+    NSData * data = [NSKeyedArchiver archivedDataWithRootObject:value requiringSecureCoding:YES error:&error];
+
+    __auto_type * controllerKey = [NSString stringWithFormat:@"%@/%@", _controllerID, key];
+    self.storage[controllerKey] = data;
+    return YES;
+}
+
+- (BOOL)controller:(MTRDeviceController *)controller
+    removeValueForKey:(NSString *)key
+        securityLevel:(MTRStorageSecurityLevel)securityLevel
+          sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Remove value for %@", controller, key);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    __auto_type * controllerKey = [NSString stringWithFormat:@"%@/%@", _controllerID, key];
+    self.storage[controllerKey] = nil;
+    return YES;
+}
+
+- (NSDictionary<NSString *, id<NSSecureCoding>> *)valuesForController:(MTRDeviceController *)controller securityLevel:(MTRStorageSecurityLevel)securityLevel sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Get all values", controller);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    NSMutableDictionary * valuesToReturn = [NSMutableDictionary dictionary];
+    for (NSString * controllerKey in self.storage) {
+        if (![controllerKey hasPrefix:[_controllerID UUIDString]]) {
+            continue;
+        }
+        __auto_type * key = [controllerKey stringByReplacingOccurrencesOfString:[_controllerID UUIDString] withString:@""];
+        valuesToReturn[key] = [self controller:controller valueForKey:key securityLevel:securityLevel sharingType:sharingType];
+    }
+
+    if (!valuesToReturn.count) {
+        return nil;
+    }
+
+    return valuesToReturn;
+}
+
+- (BOOL)controller:(MTRDeviceController *)controller storeValues:(NSDictionary<NSString *, id<NSSecureCoding>> *)values securityLevel:(MTRStorageSecurityLevel)securityLevel sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - store values", controller);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    for (NSString * controllerKey in values) {
+        if (![controllerKey hasPrefix:[_controllerID UUIDString]]) {
+            continue;
+        }
```

Comment:
I don't follow this part. These are key values coming from the framework; how would it know to prefix them with your uuid?  This does not look right.

Reply:
> I don't follow this part. These are key values coming from the framework; how would it know to prefix them with your uuid? This does not look right.
You’re right; this doesn’t make sense. It was an oversight due to copy-pasting. Thanks for catching that!